lib/db/transactions.go

```diff
@@ -7,111 +7,152 @@
+		ReadTransaction: tran,
+		keyer:           db.keyer,
+	}, nil
 }
 
 func (t readOnlyTransaction) close() {
 	t.Release()
 }
 
-func (t readOnlyTransaction) getFile(folder, device, file []byte) (protocol.FileInfo, bool) {
-	return t.getFileByKey(t.keyer.GenerateDeviceFileKey(nil, folder, device, file))
+func (t readOnlyTransaction) getFile(folder, device, file []byte) (protocol.FileInfo, bool, error) {
+	key, err := t.keyer.GenerateDeviceFileKey(nil, folder, device, file)
+	if err != nil {
+		return protocol.FileInfo{}, false, err
+	}
+	return t.getFileByKey(key)
 }
 
-func (t readOnlyTransaction) getFileByKey(key []byte) (protocol.FileInfo, bool) {
-	if f, ok := t.getFileTrunc(key, false); ok {
-		return f.(protocol.FileInfo), true
+func (t readOnlyTransaction) getFileByKey(key []byte) (protocol.FileInfo, bool, error) {
+	f, ok, err := t.getFileTrunc(key, false)
+	if err != nil {
+		return protocol.FileInfo{}, false, err
+	}
+	if !ok {
+		return protocol.FileInfo{}, false, nil
 	}
-	return protocol.FileInfo{}, false
+	return f.(protocol.FileInfo), true, nil
 }
 
-func (t readOnlyTransaction) getFileTrunc(key []byte, trunc bool) (FileIntf, bool) {
-	bs, err := t.Get(key, nil)
-	if err == leveldb.ErrNotFound {
-		return nil, false
+func (t readOnlyTransaction) getFileTrunc(key []byte, trunc bool) (FileIntf, bool, error) {
+	bs, err := t.Get(key)
+	if backend.IsNotFound(err) {
+		return nil, false, nil
 	}
 	if err != nil {
-		l.Debugln("surprise error:", err)
-		return nil, false
+		return nil, false, err
 	}
 	f, err := unmarshalTrunc(bs, trunc)
 	if err != nil {
-		l.Debugln("unmarshal error:", err)
-		return nil, false
+		return nil, false, err
 	}
-	return f, true
+	return f, true, nil
 }
 
-func (t readOnlyTransaction) getGlobal(keyBuf, folder, file []byte, truncate bool) ([]byte, FileIntf, bool) {
-	keyBuf = t.keyer.GenerateGlobalVersionKey(keyBuf, folder, file)
+func (t readOnlyTransaction) getGlobal(keyBuf, folder, file []byte, truncate bool) ([]byte, FileIntf, bool, error) {
+	var err error
+	keyBuf, err = t.keyer.GenerateGlobalVersionKey(keyBuf, folder, file)
+	if err != nil {
+		return nil, nil, false, err
+	}
 
-	bs, err := t.Get(keyBuf, nil)
+	bs, err := t.Get(keyBuf)
+	if backend.IsNotFound(err) {
+		return keyBuf, nil, false, nil
+	}
 	if err != nil {
-		return keyBuf, nil, false
+		return nil, nil, false, err
 	}
 
 	vl, ok := unmarshalVersionList(bs)
 	if !ok {
-		return keyBuf, nil, false
+		return keyBuf, nil, false, nil
 	}
 
-	keyBuf = t.keyer.GenerateDeviceFileKey(keyBuf, folder, vl.Versions[0].Device, file)
-	if fi, ok := t.getFileTrunc(keyBuf, truncate); ok {
-		return keyBuf, fi, true
+	keyBuf, err = t.keyer.GenerateDeviceFileKey(keyBuf, folder, vl.Versions[0].Device, file)
+	if err != nil {
+		return nil, nil, false, err
 	}
-
-	return keyBuf, nil, false
+	fi, ok, err := t.getFileTrunc(keyBuf, truncate)
+	if !ok {
+		return keyBuf, nil, false, nil
+	}
+	if err != nil {
+		return nil, nil, false, err
```

Comment:
Can still return `keyBuf` and thus also merge the two ifs. Sure we're probably panicking thus making reuse not really an issue, but it also doesn't hurt.

Reply:
I really think that in this error case there is no continuation possible and it's cleaner to return all the nils to indicate that all bets are off at this point. Returning the keybuf, so far as it indicates anything at all, indicates that just this little thing didn't work out quite right but please continue calling more db functions, which isn't really the thing...

Reply:
My point is that I believe the distinction (whether returning `nil` or `keyBuf`) really doesn't matter at all when `err != nil` and I consider a second, separate if branche noise that reduces readability. I don't think anyone would even look at `keyBuf` when encountering an error, even less go like "lets do something drastic if err != nil, unless keyBuf != nil, then try to recover somehow, because apparently they want me to keep using that buffer".

Reply:
OK, fair

Reply:
Good point too - might have asked Audrius for a third opinion before going through, but maybe it's just as well to stop bikeshedding at this point (sorry about that ;) ).